src/pybind/mgr/dashboard/controllers/nvmeof.py

```diff
@@ -350,6 +397,73 @@ def list(self, nqn: str):
                 NVMeoFClient.pb2.list_hosts_req(subsystem=nqn)
             )
 
+        @staticmethod
+        def determine_nqns(res, new_host_nqns):
+            to_delete_nqns = []
+            to_add_nqns = []
+
+            current_host_nqns = [host.nqn for host in res.hosts]
+
+            # All hosts are allowed but request is to add new hosts
+            if res.allow_any_host and new_host_nqns != ["*"]:
+                to_delete_nqns = ["*"]
+                to_add_nqns = new_host_nqns
+
+            # All hosts are not allowed but request is to allow all
+            elif new_host_nqns == ["*"]:
+                to_delete_nqns = current_host_nqns
+                to_add_nqns.append("*")
+
+            # Add new hosts only
+            else:
+                for current_nqn in current_host_nqns:
+                    if current_nqn not in new_host_nqns:
+                        to_delete_nqns.append(current_nqn)
+
+                for new_nqn in new_host_nqns:
+                    if new_nqn not in current_host_nqns:
+                        to_add_nqns.append(new_nqn)
+
+            return to_delete_nqns, to_add_nqns
+
+        @EndpointDoc(
+            "Allow hosts to access an NVMeoF subsystem",
+            parameters={
+                "nqn": Param(str, "NVMeoF subsystem NQN"),
+                "host_nqn": Param(str, 'NVMeoF host NQN. Use "*" to allow any host.'),
+            },
+        )
+        @empty_response
+        @handle_nvmeof_error
+        @allow_empty_body
+        def set(self, nqn: str, host_nqn: str):
+            res = NVMeoFClient().stub.list_hosts(
+                NVMeoFClient.pb2.list_hosts_req(subsystem=nqn))
+            if res.status != 0:
+                return res
+
+            new_host_nqns = host_nqn.split(',')
+            to_delete_nqns, to_add_nqns = self.determine_nqns(res, new_host_nqns)
+
+            for del_nqn in to_delete_nqns:
+                res = NVMeoFClient().stub.remove_host(
+                    NVMeoFClient.pb2.remove_host_req(subsystem_nqn=nqn, host_nqn=del_nqn)
+                )
+                if res.status != 0:
+                    return res
+                logger.info("removed host %s from subsystem %s", del_nqn, nqn)
+
+            for add_nqn in to_add_nqns:
+                res = NVMeoFClient().stub.add_host(
+                    NVMeoFClient.pb2.add_host_req(subsystem_nqn=nqn, host_nqn=add_nqn)
+                )
+                if res.status != 0:
+                    return res
+                logger.info("added host %s to subsystem %s", add_nqn, nqn)
+
+            logger.info("Updated the hosts in subsystem %s", nqn)
+            return res
```

Comment:
Same question. Do we want subsystem APIs to make so many things?

Reply:
This is the basic PUT api for hosts `PUT /api/nvmeof/subsystem/{nqn}/host/{host_nqn}`

This is needed  to update hosts per subsystem. I mean if I have 10 initiators per subsystem then instead of having 10 requests I can send one request. This will be extremely helpful from UI POV.

You may check the attached screenhost of  UI where its used.